trajoptlib/src/path/DifferentialPathBuilder.cpp

```diff
@@ -0,0 +1,214 @@
+    for (auto& obstaclePoint : obstacle.points) {
+      // First apply constraint for all but last edge
+      for (size_t bumperCornerIndex = 0;
+           bumperCornerIndex < bumperCornerCount - 1; bumperCornerIndex++) {
+        SgmtConstraint(
+            fromIndex, toIndex,
+            LinePointConstraint{_bumpers.points.at(bumperCornerIndex),
+                                _bumpers.points.at(bumperCornerIndex + 1),
+                                obstaclePoint, minDistance});
+      }
+      // apply to last edge: the edge connecting the last point to the first
+      // must have at least three points to need this
+      if (bumperCornerCount >= 3) {
+        SgmtConstraint(fromIndex, toIndex,
+                       LinePointConstraint{
+                           _bumpers.points.at(bumperCornerCount - 1),
+                           _bumpers.points.at(0), obstaclePoint, minDistance});
+      }
+    }
+
+    // obstacle edge to bumper corner constraints
+    for (auto& bumperCorner : _bumpers.points) {
+      if (obstacleCornerCount > 1) {
+        for (size_t obstacleCornerIndex = 0;
+             obstacleCornerIndex < obstacleCornerCount - 1;
+             obstacleCornerIndex++) {
+          SgmtConstraint(
+              fromIndex, toIndex,
+              PointLineConstraint{
+                  bumperCorner, obstacle.points.at(obstacleCornerIndex),
+                  obstacle.points.at(obstacleCornerIndex + 1), minDistance});
+        }
+        if (obstacleCornerCount >= 3) {
+          SgmtConstraint(
+              fromIndex, toIndex,
+              PointLineConstraint{bumperCorner,
+                                  obstacle.points.at(bumperCornerCount - 1),
+                                  obstacle.points.at(0), minDistance});
+        }
+      } else {
+        SgmtConstraint(fromIndex, toIndex,
+                       PointPointConstraint{bumperCorner, obstacle.points.at(0),
+                                            minDistance});
+      }
+    }
+  }
+}
+
+void DifferentialPathBuilder::ControlIntervalCounts(
+    std::vector<size_t>&& counts) {
+  controlIntervalCounts = std::move(counts);
+}
+
+const std::vector<size_t>& DifferentialPathBuilder::GetControlIntervalCounts()
+    const {
+  return controlIntervalCounts;
+}
+
+DifferentialSolution DifferentialPathBuilder::CalculateInitialGuess() const {
+  return GenerateLinearInitialGuess<DifferentialSolution>(
+      initialGuessPoints, controlIntervalCounts);
+}
+
+void DifferentialPathBuilder::NewWpts(size_t finalIndex) {
+  int64_t targetIndex = finalIndex;
+  int64_t greatestIndex = path.waypoints.size() - 1;
+  if (targetIndex > greatestIndex) {
+    for (int64_t i = greatestIndex + 1; i <= targetIndex; ++i) {
+      path.waypoints.emplace_back();
+      initialGuessPoints.emplace_back(std::vector<Pose2d>{{0.0, 0.0, {0.0}}});
+      if (i != 0) {
+        controlIntervalCounts.push_back(40);
+      }
+    }
+  }
+}
```

Comment:
You can deduplicate the common stuff between swerve and differential with either inheritance or composition.

Reply:
Should this be done for the functions inlined functions in SwerveTrajectoryGenerator and DifferentialTrajectoryGenerator